### ISSUE TITLE
StreamField adds duplicate blocks if user filters by block name

### ISSUE THREAD

**Issue by chosak** [wagtail/wagtail#10057]
### Issue Summary

When adding a block to a StreamField, filtering the list of blocks by name causes two blocks to be inserted instead of only one.

Correct behavior: clicking the block name "Text" inserts one block:

![right](https://user-images.githubusercontent.com/654645/218098138-f3490b65-c7d6-4445-ab83-9c87fb643a37.gif)

Incorrect behavior: typing the letter "t" to first filter block names and then clicking on "Text" unexpectedly inserts two blocks:

![wrong](https://user-images.githubusercontent.com/654645/218098159-44f2522b-5edd-458b-8703-a14ca2dee3a3.gif)

### Steps to Reproduce

1. Start a new project with `wagtail start myproject`
2. Edit models.py as follows:

```py
from wagtail import blocks, fields
from wagtail.models import Page
from wagtail.admin.panels import FieldPanel

class HomePage(Page):
    body = fields.StreamField([("text", blocks.TextBlock())], blank=True, use_json_field=True)

    content_panels = Page.content_panels + [FieldPanel("body")]
```

3. `./manage.py makemigrations && ./manage.py migrate && ./manage.py createsuperuser`
4. Edit the existing HomePage at http://localhost:8000/admin/pages/3/edit/.
5. Try behavior above: type a letter to filter the list of available blocks and then click on the block name. Two blocks get entered instead of one.

This seems to happen regardless of the block type. This duplication behavior also happens even if the StreamField is set with `max_num=1`, although this may be a separate issue covered by #9513.

-   I have confirmed that this issue can be reproduced as described on a fresh Wagtail project: yes

### Technical details

-   Python version: 3.11.1
-   Django version: 4.1.6
-   Wagtail version: 4.2
-   Browser version: Chrome 103

**Comment by arunimaChintu** [comment]
I want to work on this issue.Can you please guide me where and how to start?

**Comment by lb-** [comment]
@arunimaChintu - this hasn't been confirmed yet. Feel free to help triage this by trying to reproduce it in our other supported browsers and then suggest where the problem may be in the code.

**Comment by arunimaChintu** [comment]
is the code written in python?

**Comment by wojcikmat** [comment]
It happened to me as well. It's pretty annoying tbh. You don't even need to type any letter into search filter, just click on it and then add a new block. It is added twice.

https://user-images.githubusercontent.com/26712858/219355358-6701dc48-c7ce-41bd-af7f-c47a9b88d104.mov

**Comment by chosak** [comment]
@wojcikmat thank you for confirming and for the additional failure case.

@arunimaChintu no, this behavior is likely related to the frontend client code which is written in TypeScript. My guess would be somewhere in the new [ComboBox](https://github.com/wagtail/wagtail/blob/357edf2914bfb7b1265011b1009f0d57ff9824f1/client/src/components/ComboBox/ComboBox.tsx) which was introduced in 4.2 in #9908.

**Comment by thibaudcolas** [comment]
Thank you @chosak, yes there’s most likely a race condition between the click event handling on each item, and the keyboard event handling on the whole autocomplete. Due to this component being reused in rich text, it has to have rather complex keyboard focus handling which is likely causing this.

**Comment by deepampriyadarshi** [comment]
Hi @chosak, @thibaudcolas. Sorry I didn't realize that this got assigned to someone, since I was not subscribed to this thread. Nevertheless, I have been working on this issue since last week and I believe that my findings will speed up the release process, if not get merged. 

There's no race condition happening. It's simply improper event handling and then calling the same function 2 times for a single select operation, with 2 different object argument, where one being a reduced version of other. 

> Then when the blur happens, because that state change isn't reflected [here](https://github.com/downshift-js/downshift/blob/7540e3d925ea9eeddf9fe3780b5b4117563e1b1c/src/hooks/utils.js#L34), a second selection change gets fired.

There's no second selection getting fired. As per downshift [docs](https://github.com/downshift-js/downshift/tree/master/src/hooks/useCombobox#onselecteditemchange) selection can be done in three ways:
![image](https://user-images.githubusercontent.com/65447610/221207701-2cb3ee8f-485a-48c4-94bc-16c3e08e6f21.png)

So click and input box focus change will always be simultaneous for a single selection action. The state change on `__input_blur__` is an inbuilt and a very desired functionality of comboxes used in form fill ups where power users don't use mouse to fill the form. Simply type in the prefix of the word => autocomplete filters out and highlights the most appropriate item => you select that item by moving to next form field(pressing tab), using combination of arrows and enter key or by mouse click. **Here also 2 simultaneous events may occur, mouse click and focus change of input area.** But that will end in setting the form field value twice, which is not a bad UX unless your final set value is not the same.

DownShift provides api to filter out and suppress unwanted events. The default click even of downshift has been overridden by event [this](https://github.com/thibaudcolas/wagtail/blob/a7df17a50707413655f22ea62af170d8fb49b75c/client/src/components/ComboBox/ComboBox.tsx#L203-L209) but blur event is in default and active.

> Then when the blur happens, because that state change isn't reflected [here](https://github.com/downshift-js/downshift/blob/7540e3d925ea9eeddf9fe3780b5b4117563e1b1c/src/hooks/utils.js#L34)

It is reflected but only by the blur event and not by click event, since it has been overridden to call an external class function [here](https://github.com/wagtail/wagtail/blob/fd73fc29058d1fa6b252e5348341365a6c6d98a9/client/src/components/StreamField/blocks/StreamBlock.js#L126-L147). [This](https://github.com/thibaudcolas/wagtail/blob/a7df17a50707413655f22ea62af170d8fb49b75c/client/src/components/ComboBox/ComboBox.tsx#L203-L209) overridden change or sets the state of ComboBox, otherwise it would have been a rave condition. It's a simple function call

**Comment by thibaudcolas** [comment]
Thank you all! It’s great to see so many people involved in identifying a bug. I’ve merged @deepampriyadarshi’s fix in the upcoming Wagtail 5.0, and also for a v4.2.1 patch release, which we currently plan to release next week. This was introduced in v4.2 so there won’t be a need for a v4.1 patch.

---

A few notes for future reference.

An alternative fix for this would have been to dismiss the block picker immediately after selection, like we do when it appears in rich text. 

Here: https://github.com/wagtail/wagtail/blame/90585c9bba722ec254b5f08a75f9b03dedb6c1c7/client/src/components/StreamField/blocks/StreamBlock.js#L92

Add:

```js
duration: [300, 0],
```

This immediately fires Tippy’s `onHide` with no animation, preventing any duplicate selections.

Technically adding a block on `blur` is more correct for a `combobox` – if there is a selected item in the `listbox`, then dismissing the listbox with keyboard navigation results in the value being selected (like a `<select>`). This is probably unexpected for most people in our implementation, because the filtering `textbox` and its `listbox` are always concurrently visible.
